src/main/java/org/tikv/common/util/ChannelFactory.java

```diff
@@ -320,8 +239,46 @@ public ChannelFactory(
     this.keepaliveTimeout = keepaliveTimeout;
     this.idleTimeout = idleTimeout;
     this.connRecycleTime = connRecycleTime;
-    this.certContext =
-        new JksContext(path, jksKeyPath, jksKeyPassword, jksTrustPath, jksTrustPassword);
+    this.certContext = new JksContext(jksKeyPath, jksKeyPassword, jksTrustPath, jksTrustPassword);
+
+    File jksKey = new File(jksKeyPath);
+    File jksTrust = new File(jksTrustPath);
+    if (certReloadInterval > 0) {
+      onCertChange();
+      this.certWatcher =
+          new CertWatcher(
+              certReloadInterval, ImmutableList.of(jksKey, jksTrust), this::onCertChange);
+    } else {
+      this.certWatcher = null;
+    }
+  }
+
+  private void onCertChange() {
+    try {
+      SslContextBuilder newBuilder = certContext.createSslContextBuilder();
+      lock.writeLock().lock();
+      sslContextBuilder.set(newBuilder);
+
+      List<ManagedChannel> pending = new ArrayList<>(connPool.values());
+      recycler.schedule(() -> cleanExpireConn(pending), connRecycleTime, TimeUnit.SECONDS);
```

Comment:
Since recycler is single-thread and prev task is still running even if cleaning was done.
So the interval of the current task is n*connRecycleTime(n is the number of tasks in the recycler).
https://blog.csdn.net/weixin_38106322/article/details/104494106